### ISSUE TITLE
auto-shrink not triggered

### ISSUE THREAD

**Issue by slingamn** [tidwall/buntdb#77]
This may be a duplicate of #37. I'm running v1.2.4. I have a database with only 37 actual entries (the workload is rapid repeated updates to a small set of keys). I'm using the default config, but autoshrink is not triggering despite the database reaching 171MB in size.

I added some prints here:

https://github.com/tidwall/buntdb/blob/v1.2.4/buntdb.go#L566

and both `aofsz` and `db.lastaofsz` are set to the file size (currently 178954614), so `shrink` evaluates to false. I'm not sure whether the condition is buggy, or whether `db.lastaofsz` is being maintained incorrectly?

Thanks for your time.

**Comment by tidwall** [comment]
I did some test on my side and I could not reproduce the problem. 

The `lastaofsz` value is only set in one of two places, 1. when the database is opened, 2. after a shrink is completed.

If it's not being triggered at all then that indicates that the lastaofsz was set when the database opened. 

Some possible scenarios that might cause this issue:

- The aof file grew to a large size and was then was closed before a shrink operation was completed.
- The AutoShink was turned off at some point and then the database was closed and reopened with AutoShink on.

With the default AutoShink options; If the aof file is 171MB when the database is opened, then eventually the aof file will get shrunk when it reaches 342MB.

Optionally, you can call `db.Shink()` manually after you open the database.

**Comment by slingamn** [comment]
Thanks for looking into this.

It's not possible for `AutoShrinkDisabled` to have been set to true in this codebase. I think what happened is that we got very unlucky with the pattern of process restarts: we restarted multiple times, the first time before we hit the 32MB threshold, then subsequent times before the file size had doubled from its size at load.

It seems like the issue is that `(*DB).load()` assumes that the file size on load is the "shrunk" size? So an effective mitigation would be either to persist the true "shrunk" size (either 0 in the case where the database was never shrunk, or the size at last shrink), or possibly to have `load()` estimate the shrunk size based on the actual data size of the btree.
